segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/PravegaRequestProcessor.java

```diff
@@ -704,11 +710,152 @@ public void readTable(final WireCommands.ReadTable readTable) {
                                               .map(k -> new ByteArraySegment(k.getData().array()))
                                               .collect(Collectors.toList());
         tableStore.get(segment, keys, TIMEOUT)
-                  .thenAccept(values -> connection.send(new WireCommands.TableRead(readTable.getRequestId(), segment, getTableEntries(keys, values))))
+                  .thenAccept(values -> connection.send(new WireCommands.TableRead(readTable.getRequestId(), segment,
+                                                                                   getTableEntriesCommand(keys, values))))
                   .exceptionally(e -> handleException(readTable.getRequestId(), segment, operation, e));
     }
 
-    private WireCommands.TableEntries getTableEntries(final List<ArrayView> inputKeys, final List<TableEntry> resultEntries) {
+    @Override
+    public void readTableKeys(WireCommands.ReadTableKeys readTableKeys) {
+        final String segment = readTableKeys.getSegment();
+        final String operation = "readTableKeys";
+
+        if (!verifyToken(segment, readTableKeys.getRequestId(), readTableKeys.getDelegationToken(), operation)) {
+            return;
+        }
+
+        log.info(readTableKeys.getRequestId(), "Fetching keys from {}.", readTableKeys);
+
+        int suggestedKeyCount = readTableKeys.getSuggestedKeyCount();
+        ByteBuf token = readTableKeys.getContinuationToken();
+
+        byte[] state = null;
+        if (!token.equals(EMPTY_BUFFER)) {
+            state = token.array();
+        }
+
+        final AtomicInteger msgSize = new AtomicInteger(0);
+        final AtomicReference<ByteBuf> continuationToken = new AtomicReference<>(EMPTY_BUFFER);
+        final List<TableKey> keys = synchronizedList(new ArrayList<>());
+
+        tableStore.keyIterator(segment, state, TIMEOUT)
+                  .thenCompose(itr -> itr.collectUntil(
+                          e -> {
+                              Collection<TableKey> tableKeys = e.getEntries();
+                              ArrayView lastState = e.getState();
+
+                              // Store all tableKeys.
+                              keys.addAll(tableKeys);
+                              // update the continuation token.
+                              continuationToken.set(wrappedBuffer(lastState.array(), lastState.arrayOffset(), lastState.getLength()));
+                              // Update msgSize.
+                              msgSize.addAndGet(getTableKeyBytes(segment, tableKeys, lastState.getLength()));
+                          },
+                          e -> keys.size() < suggestedKeyCount && msgSize.get() < MAX_READ_SIZE))
+                  .thenAccept(v -> {
+                      log.debug(readTableKeys.getRequestId(), "{} keys obtained for ReadTableKeys request.", keys.size());
+                      List<WireCommands.TableKey> wireCommandKeys = keys.stream()
```

Comment:
Unfortunately, Java decided to be totally dumb. This doesn't work. The stream method on a synchronized list is not synchronized. (I know)

Reply:
corrected it.